pkg/util/watch/delete.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watch
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// These log fields
+const (
+	logFieldObjectRef  = "object"
+	logFieldObjectKind = "objectKind"
+)
+
+// DeleteAndWait deletes the object (if not already deleted) and blocks until it
+// is fully deleted (NotFound, not just deleteTimestamp). This allows for any
+// finalizers to complete.
+//
+// Exits early if the object is already deleted or doesn't exist.
+//
+// The Object contents will be replaced with updated state for each new event
+// received while waiting for deletion.
+//
+// If you want to time out the full function, use a context.WithTimeout.
```

Comment:
It seems reconcileTimeout still takes effect even if a context.WithTimeout is used. Right?

Reply:
Correct.